sklearn/linear_model/coordinate_descent.py

```diff
@@ -692,7 +690,7 @@ def fit(self, X, y, check_input=True):
         if self.selection not in ['cyclic', 'random']:
             raise ValueError("selection should be either random or cyclic.")
 
-        if not self.warm_start or self.coef_ is None:
+        if not self.warm_start or getattr(self, "coef_", None) is None:
```

Comment:
https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/decomposition/dict_learning.py#L125 is the problem. It sets `coef_` directly before fitting. You can either fix that to only set `coef_` if `init is not None`, or continue with the approach used here.